drv/cosmo-hf/src/hf.rs

```diff
@@ -218,15 +364,54 @@ impl idl::InOrderHostFlashImpl for ServerImpl {
         &mut self,
         _: &RecvMessage,
     ) -> Result<HfPersistentData, RequestError<HfError>> {
-        todo!()
+        self.get_persistent_data().map_err(RequestError::from)
     }
 
     fn write_persistent_data(
         &mut self,
         _: &RecvMessage,
-        _dev_select: HfDevSelect,
+        dev_select: HfDevSelect,
     ) -> Result<(), RequestError<HfError>> {
-        todo!()
+        let data = HfPersistentData { dev_select };
+
+        let prev_slot = self.dev;
+
+        // Scan both slots for persistent data
+        self.dev = !prev_slot;
+        let (a_data, a_next) = self.persistent_data_scan();
+
+        self.dev = prev_slot;
+        let (b_data, b_next) = self.persistent_data_scan();
```

Comment:
aaaaaand my comments about the toggling of `self.dev` apply to this function as well, of course.